File: apps/frontend/types/storage.ts
// Storage System TypeScript Interfaces
// Based on 17_storage_system.md and 18_storage_system_manager.md

import type {
  StorageUnitType, 
  StorageUnitStatus, 
  StorageLocationType, 
  StorageLocationStatus, 
  PaymentStatus, 
  BookingStatus, 
  StorageUserRole,
  Currency,
  BillingCycle,
  PaymentMethod,
  DiscountType,
  DiscountPolicyType,
  AccessType,
  SecurityFeatureType,
  AlertSeverity,
  AlertType,
  MaintenanceType,
  TemperatureUnit,
  BookingSource,
  BookingUrgency,
  ChangeType,
  ApiResponseStatus
} from './enums';


// Storage Unit Interfaces
export interface StorageSize {
  width: number;
  length: number;
  height: number;
  unit: 'feet' | 'meters';
}

export interface MapPosition {
  x: number;
  y: number;
  rotation: number;
  gridPosition: GridPosition;
}

export interface GridPosition {
  row: number;
  column: number;
}

export interface StoragePricing {
  basePrice: number;
  currency: Currency;
  billingCycle: BillingCycle;
  discounts: Discount[];
}

export interface Discount {
  id: string;
  type: DiscountType;
  value: number;
  description: string;
  validFrom: Date;
  validTo: Date;
}

export interface CustomerInfo {
  id: string;
  name: string;
  email: string;
  phone: string;
  startDate: Date;
  endDate?: Date;
  paymentStatus: PaymentStatus;
}

export interface StorageFeature {
  id: string;
  name: string;
  description: string;
  enabled: boolean;
}

export interface MaintenanceRecord {
  id: string;
  date: Date;
  type: MaintenanceType;
  description: string;
  cost: number;
  performedBy: string;
  notes: string;
}

export interface StorageUnit {
  id: string;
  locationId: string;
  type: StorageUnitType;
  size: StorageSize;
  position: MapPosition;
  status: StorageUnitStatus;
  pricing: StoragePricing;
  customer?: CustomerInfo;
  features: StorageFeature[];
  maintenanceHistory: MaintenanceRecord[];
  createdAt: Date;
  updatedAt: Date;
}

// Storage Location Interfaces
export interface StorageLocation {
  id: string;
  name: string;
  companyId: string;
  type: StorageLocationType;
  status: StorageLocationStatus;
  address: {
    street: string;
    city: string;
    province: string;
    country: string;
  };
  coordinates: {
    latitude: number;
    longitude: number;
  };
  contact: {
    manager: string;
    phone: string;
    email: string;
    emergency: string;
  };
  hours: {
    monday: { open: string; close: string; closed: boolean };
    tuesday: { open: string; close: string; closed: boolean };
    wednesday: { open: string; close: string; closed: boolean };
    thursday: { open: string; close: string; closed: boolean };
    friday: { open: string; close: string; closed: boolean };
    saturday: { open: string; close: string; closed: boolean };
    sunday: { open: string; close: string; closed: boolean };
    timezone: string;
  };
  storage: {
    types: StorageUnitType[];
    totalCapacity: number;
    availableCapacity: number;
    layout: StorageLayout;
    security: SecurityFeatures[];
  };
  policies: {
    accessHours: string;
    securityRequirements: string[];
    maintenanceSchedule: string;
    emergencyProcedures: string;
  };
  pricing: {
    baseRates: StoragePricing;
    discounts: DiscountPolicy[];
    paymentTerms: PaymentTerms;
    lateFees: LateFeePolicy;
  };
}

export interface StorageLayout {
  width: number;
  length: number;
  height: number;
  unit: 'feet' | 'meters';
  gridSize: number;
  accessPaths: AccessPath[];
}

export interface AccessPath {
  id: string;
  startPoint: { x: number; y: number };
  endPoint: { x: number; y: number };
  width: number;
  type: 'MAIN' | 'SECONDARY' | 'EMERGENCY';
}

export interface SecurityFeatures {
  id: string;
  name: string;
  type: SecurityFeatureType;
  enabled: boolean;
  description: string;
}

export interface DiscountPolicy {
  id: string;
  name: string;
  type: DiscountPolicyType;
  value: number;
  conditions: string[];
  validFrom: Date;
  validTo: Date;
}

export interface PaymentTerms {
  dueDate: number; // days from invoice
  lateFeePercentage: number;
  gracePeriod: number;
  autoRenewal: boolean;
}

export interface LateFeePolicy {
  percentage: number;
  minimumAmount: number;
  maximumAmount: number;
  gracePeriod: number;
}

// Storage Zone Interface
export interface StorageZone {
  id: string;
  locationId: string;
  name: string;
  type: StorageUnitType | 'MIXED';
  status: StorageLocationStatus;
  layout: {
    width: number;
    length: number;
    height: number;
    unit: 'feet' | 'meters';
    gridSize: number;
    accessPaths: AccessPath[];
  };
  capacity: {
    totalUnits: number;
    availableUnits: number;
    reservedUnits: number;
    maintenanceUnits: number;
    utilizationRate: number;
  };
  security: {
    accessType: AccessType;
    surveillance: boolean;
    lighting: boolean;
    fencing: boolean;
    accessLogs: boolean;
  };
  environment: {
    climateControlled: boolean;
    temperature: {
      min: number;
      max: number;
      unit: TemperatureUnit;
    };
    humidity: {
      min: number;
      max: number;
    };
    ventilation: boolean;
  };
}

// Storage Map Interface
export interface StorageMap {
  locationId: string;
  mapConfig: MapConfiguration;
  storageUnits: StorageUnit[];
  zones: StorageZone[];
  capacity: StorageCapacity;
  realTimeData: RealTimeStorageData;
}

export interface MapConfiguration {
  width: number;
  height: number;
  gridSize: number;
  snapToGrid: boolean;
  collisionDetection: boolean;
  undoRedo: boolean;
}

export interface StorageCapacity {
  total: number;
  available: number;
  occupied: number;
  reserved: number;
  maintenance: number;
  utilizationRate: number;
}

export interface RealTimeStorageData {
  lastUpdated: Date;
  activeUsers: number;
  recentChanges: StorageChange[];
  alerts: StorageAlert[];
}

export interface StorageChange {
  id: string;
  timestamp: Date;
  userId: string;
  unitId: string;
  changeType: ChangeType;
  oldValue: any;
  newValue: any;
}

export interface StorageAlert {
  id: string;
  type: AlertType;
  severity: AlertSeverity;
  message: string;
  timestamp: Date;
  resolved: boolean;
}

// Drag and Drop Configuration
export interface DragDropConfig {
  draggable: boolean;
  resizable: boolean;
  rotatable: boolean;
  snapToGrid: boolean;
  collisionDetection: boolean;
  undoRedo: boolean;
}

// Analytics Interfaces
export interface StorageAnalytics {
  totalUnits: number;
  occupiedUnits: number;
  availableUnits: number;
  utilizationRate: number;
  revenuePerUnit: number;
  totalRevenue: number;
  averageOccupancy: number;
  turnoverRate: number;
}

export interface OperationalKPIs {
  utilization: {
    overallUtilization: number;
    unitUtilization: number;
    zoneUtilization: number;
    seasonalUtilization: number;
  };
  customerService: {
    responseTime: number;
    resolutionTime: number;
    customerSatisfaction: number;
    complaintRate: number;
  };
  maintenance: {
    preventiveMaintenance: number;
    emergencyMaintenance: number;
    maintenanceCost: number;
    downtime: number;
  };
  security: {
    securityIncidents: number;
    unauthorizedAccess: number;
    securityResponseTime: number;
    complianceScore: number;
  };
}

export interface FinancialKPIs {
  revenue: {
    totalRevenue: number;
    revenuePerUnit: number;
    revenueGrowth: number;
    averageOccupancy: number;
  };
  costs: {
    operationalCosts: number;
    maintenanceCosts: number;
    securityCosts: number;
    staffCosts: number;
  };
  profitability: {
    grossMargin: number;
    netMargin: number;
    returnOnInvestment: number;
    breakEvenPoint: number;
  };
  billing: {
    paymentOnTime: number;
    latePayments: number;
    collectionRate: number;
    averagePaymentTime: number;
  };
}

// Booking Interfaces
export interface StorageBooking {
  id: string;
  unitId: string;
  customerId: string;
  startDate: Date;
  endDate?: Date;
  totalCost: number;
  paymentStatus: PaymentStatus;
  status: BookingStatus;
  createdAt: Date;
  updatedAt: Date;
}

export interface BookingWorkflow {
  step1: {
    customerInquiry: {
      source: BookingSource;
      requirements: StorageRequirements;
      timeline: BookingTimeline;
    };
  };
  step2: {
    availabilityCheck: {
      location: string;
      unitType: StorageUnitType;
      size: StorageSize;
      duration: number;
      startDate: Date;
    };
  };
  step3: {
    pricingCalculation: {
      basePrice: number;
      duration: number;
      discounts: Discount[];
      taxes: number;
      totalPrice: number;
    };
  };
  step4: {
    customerRegistration: {
      personalInfo: CustomerInfo;
      contactInfo: ContactDetails;
      paymentInfo: StoragePaymentMethod;
      termsAcceptance: boolean;
    };
  };
  step5: {
    bookingConfirmation: {
      bookingId: string;
      unitAssignment: StorageUnit;
      accessInstructions: AccessInfo;
      paymentConfirmation: PaymentConfirmation;
    };
  };
  step6: {
    customerOnboarding: {
      welcomeEmail: boolean;
      accessCodeIssuance: boolean;
      facilityOrientation: boolean;
      supportContact: boolean;
    };
  };
}

export interface StorageRequirements {
  size: StorageSize;
  type: StorageUnitType;
  duration: number;
  specialNeeds: string[];
  budget: number;
}

export interface BookingTimeline {
  startDate: Date;
  endDate?: Date;
  flexible: boolean;
  urgency: BookingUrgency;
}

export interface ContactDetails {
  name: string;
  email: string;
  phone: string;
  address: string;
}

export interface StoragePaymentMethod {
  type: PaymentMethod;
  details: any;
  autoRenewal: boolean;
}

export interface AccessInfo {
  accessCode: string;
  accessHours: string;
  facilityRules: string[];
  emergencyContact: string;
}

export interface PaymentConfirmation {
  transactionId: string;
  amount: number;
  status: ApiResponseStatus;
  receipt: string;
}

// User Permissions Interface
export interface UserPermissions {
  system: {
    configuration: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    maintenance: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    backups: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  companies: {
    create: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    read: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    update: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    delete: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  locations: {
    create: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    read: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    update: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    delete: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  storage_units: {
    create: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    read: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    update: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    delete: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    layout: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  customers: {
    create: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    read: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    update: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    delete: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  bookings: {
    create: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    read: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    update: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    cancel: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  analytics: {
    system_wide: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    financial: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    operational: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
  audit: {
    logs: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    compliance: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
    reports: 'FULL_ACCESS' | 'READ_ONLY' | 'NO_ACCESS';
  };
}

export interface StorageUser {
  id: string;
  username: string;
  email: string;
  role: StorageUserRole;
  permissions: UserPermissions;
  companyId?: string;
  locationId?: string;
  isActive: boolean;
  lastLogin: Date;
  createdAt: Date;
  updatedAt: Date;
}

// API Response Types
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// Filter Interfaces
export interface StorageFilter {
  locationId?: string;
  type?: StorageUnitType;
  status?: StorageUnitStatus;
  size?: StorageSize;
  priceRange?: {
    min: number;
    max: number;
  };
  availability?: {
    startDate: Date;
    endDate: Date;
  };
}

export interface LocationFilter {
  companyId?: string;
  type?: StorageLocationType;
  status?: StorageLocationStatus;
  province?: string;
  hasStorage?: boolean;
} 

File: apps/api/routes/storage.py
"""
Storage System API Routes
Handles storage location, unit, and booking management
"""

from fastapi import APIRouter, HTTPException, Depends, Query
from typing import List, Optional
from datetime import datetime, date
import json

router = APIRouter(prefix="/storage", tags=["Storage"])

# ===== STORAGE LOCATIONS =====

@router.get("/locations")
async def get_storage_locations(
    company_id: Optional[str] = Query(None, description="Filter by company ID"),
    type: Optional[str] = Query(None, description="Filter by location type"),
    status: Optional[str] = Query(None, description="Filter by status"),
    province: Optional[str] = Query(None, description="Filter by province"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(10, ge=1, le=100, description="Items per page")
):
    """Get storage locations with filtering and pagination"""
    try:
        # TODO: Implement database query
        # Mock response for now
        return {
            "success": True,
            "data": {
                "locations": [
                    {
                        "id": "loc1",
                        "name": "Toronto Storage Facility",
                        "companyId": "company1",
                        "type": "CORPORATE",
                        "status": "ACTIVE",
                        "address": {
                            "street": "123 Storage Ave",
                            "city": "Toronto",
                            "province": "ON",
                            "country": "Canada"
                        },
                        "coordinates": {
                            "latitude": 43.6532,
                            "longitude": -79.3832
                        },
                        "contact": {
                            "manager": "John Smith",
                            "phone": "+1-416-555-0123",
                            "email": "toronto@lgm.com",
                            "emergency": "+1-416-555-9999"
                        },
                        "createdAt": "2024-01-01T00:00:00Z",
                        "updatedAt": "2024-01-15T00:00:00Z"
                    }
                ],
                "pagination": {
                    "page": page,
                    "limit": limit,
                    "total": 1,
                    "totalPages": 1
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage locations: {str(e)}")

@router.post("/locations")
async def create_storage_location(location_data: dict):
    """Create a new storage location"""
    try:
        # TODO: Implement database creation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "id": f"loc_{int(datetime.now().timestamp())}",
                "name": location_data.get("name"),
                "companyId": location_data.get("companyId"),
                "type": location_data.get("type"),
                "status": "ACTIVE",
                "createdAt": datetime.now().isoformat(),
                "updatedAt": datetime.now().isoformat()
            },
            "message": "Storage location created successfully"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create storage location: {str(e)}")

@router.get("/locations/{location_id}")
async def get_storage_location(location_id: str):
    """Get a specific storage location by ID"""
    try:
        # TODO: Implement database query
        # Mock response for now
        return {
            "success": True,
            "data": {
                "id": location_id,
                "name": "Toronto Storage Facility",
                "companyId": "company1",
                "type": "CORPORATE",
                "status": "ACTIVE",
                "address": {
                    "street": "123 Storage Ave",
                    "city": "Toronto",
                    "province": "ON",
                    "country": "Canada"
                },
                "coordinates": {
                    "latitude": 43.6532,
                    "longitude": -79.3832
                },
                "contact": {
                    "manager": "John Smith",
                    "phone": "+1-416-555-0123",
                    "email": "toronto@lgm.com",
                    "emergency": "+1-416-555-9999"
                },
                "hours": {
                    "monday": {"open": "08:00", "close": "18:00", "closed": False},
                    "tuesday": {"open": "08:00", "close": "18:00", "closed": False},
                    "wednesday": {"open": "08:00", "close": "18:00", "closed": False},
                    "thursday": {"open": "08:00", "close": "18:00", "closed": False},
                    "friday": {"open": "08:00", "close": "18:00", "closed": False},
                    "saturday": {"open": "09:00", "close": "17:00", "closed": False},
                    "sunday": {"open": "10:00", "close": "16:00", "closed": False},
                    "timezone": "America/Toronto"
                },
                "storage": {
                    "types": ["POD", "LOCKER"],
                    "totalCapacity": 500,
                    "availableCapacity": 350
                },
                "createdAt": "2024-01-01T00:00:00Z",
                "updatedAt": "2024-01-15T00:00:00Z"
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage location: {str(e)}")

# ===== STORAGE UNITS =====

@router.get("/units")
async def get_storage_units(
    location_id: Optional[str] = Query(None, description="Filter by location ID"),
    type: Optional[str] = Query(None, description="Filter by unit type"),
    status: Optional[str] = Query(None, description="Filter by status"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(10, ge=1, le=100, description="Items per page")
):
    """Get storage units with filtering and pagination"""
    try:
        # TODO: Implement database query
        # Mock response for now
        return {
            "success": True,
            "data": {
                "units": [
                    {
                        "id": "unit1",
                        "locationId": "loc1",
                        "type": "POD",
                        "size": {
                            "width": 10,
                            "length": 10,
                            "height": 8,
                            "unit": "feet"
                        },
                        "position": {
                            "x": 100,
                            "y": 200,
                            "rotation": 0,
                            "gridPosition": {"row": 1, "column": 1}
                        },
                        "status": "AVAILABLE",
                        "pricing": {
                            "basePrice": 150.00,
                            "currency": "CAD",
                            "billingCycle": "MONTHLY"
                        },
                        "features": [],
                        "maintenanceHistory": [],
                        "createdAt": "2024-01-01T00:00:00Z",
                        "updatedAt": "2024-01-15T00:00:00Z"
                    }
                ],
                "pagination": {
                    "page": page,
                    "limit": limit,
                    "total": 1,
                    "totalPages": 1
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage units: {str(e)}")

@router.post("/units")
async def create_storage_unit(unit_data: dict):
    """Create a new storage unit"""
    try:
        # TODO: Implement database creation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "id": f"unit_{int(datetime.now().timestamp())}",
                "locationId": unit_data.get("locationId"),
                "type": unit_data.get("type"),
                "size": unit_data.get("size"),
                "position": unit_data.get("position"),
                "status": "AVAILABLE",
                "pricing": unit_data.get("pricing"),
                "features": unit_data.get("features", []),
                "maintenanceHistory": [],
                "createdAt": datetime.now().isoformat(),
                "updatedAt": datetime.now().isoformat()
            },
            "message": "Storage unit created successfully"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create storage unit: {str(e)}")

@router.put("/units/{unit_id}")
async def update_storage_unit(unit_id: str, updates: dict):
    """Update a storage unit"""
    try:
        # TODO: Implement database update
        # Mock response for now
        return {
            "success": True,
            "data": {
                "id": unit_id,
                "updatedAt": datetime.now().isoformat()
            },
            "message": "Storage unit updated successfully"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update storage unit: {str(e)}")

# ===== STORAGE BOOKINGS =====

@router.get("/bookings")
async def get_storage_bookings(
    unit_id: Optional[str] = Query(None, description="Filter by unit ID"),
    customer_id: Optional[str] = Query(None, description="Filter by customer ID"),
    status: Optional[str] = Query(None, description="Filter by booking status"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(10, ge=1, le=100, description="Items per page")
):
    """Get storage bookings with filtering and pagination"""
    try:
        # TODO: Implement database query
        # Mock response for now
        return {
            "success": True,
            "data": {
                "bookings": [
                    {
                        "id": "booking1",
                        "unitId": "unit1",
                        "customerId": "customer1",
                        "startDate": "2024-01-15T00:00:00Z",
                        "endDate": "2024-02-15T00:00:00Z",
                        "totalCost": 150.00,
                        "paymentStatus": "PAID",
                        "status": "ACTIVE",
                        "createdAt": "2024-01-10T00:00:00Z",
                        "updatedAt": "2024-01-15T00:00:00Z"
                    }
                ],
                "pagination": {
                    "page": page,
                    "limit": limit,
                    "total": 1,
                    "totalPages": 1
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage bookings: {str(e)}")

@router.post("/bookings")
async def create_storage_booking(booking_data: dict):
    """Create a new storage booking"""
    try:
        # TODO: Implement database creation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "id": f"booking_{int(datetime.now().timestamp())}",
                "unitId": booking_data.get("unitId"),
                "customerId": booking_data.get("customerId"),
                "startDate": booking_data.get("startDate"),
                "endDate": booking_data.get("endDate"),
                "totalCost": booking_data.get("totalCost"),
                "paymentStatus": "PENDING",
                "status": "ACTIVE",
                "createdAt": datetime.now().isoformat(),
                "updatedAt": datetime.now().isoformat()
            },
            "message": "Storage booking created successfully"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create storage booking: {str(e)}")

# ===== STORAGE ANALYTICS =====

@router.get("/analytics/{location_id}")
async def get_storage_analytics(location_id: str):
    """Get storage analytics for a specific location"""
    try:
        # TODO: Implement analytics calculation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "totalUnits": 100,
                "occupiedUnits": 75,
                "availableUnits": 25,
                "utilizationRate": 75.0,
                "revenuePerUnit": 150.00,
                "totalRevenue": 11250.00,
                "averageOccupancy": 85.0,
                "turnoverRate": 12.5
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage analytics: {str(e)}")

@router.get("/analytics/operational/{location_id}")
async def get_operational_kpis(location_id: str):
    """Get operational KPIs for a storage location"""
    try:
        # TODO: Implement KPI calculation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "utilization": {
                    "overallUtilization": 75.0,
                    "unitUtilization": 80.0,
                    "zoneUtilization": 70.0,
                    "seasonalUtilization": 85.0
                },
                "customerService": {
                    "responseTime": 2.5,
                    "resolutionTime": 24.0,
                    "customerSatisfaction": 4.5,
                    "complaintRate": 2.0
                },
                "maintenance": {
                    "preventiveMaintenance": 95.0,
                    "emergencyMaintenance": 5.0,
                    "maintenanceCost": 2500.00,
                    "downtime": 0.5
                },
                "security": {
                    "securityIncidents": 0,
                    "unauthorizedAccess": 0,
                    "securityResponseTime": 5.0,
                    "complianceScore": 98.0
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch operational KPIs: {str(e)}")

@router.get("/analytics/financial/{location_id}")
async def get_financial_kpis(location_id: str):
    """Get financial KPIs for a storage location"""
    try:
        # TODO: Implement financial KPI calculation
        # Mock response for now
        return {
            "success": True,
            "data": {
                "revenue": {
                    "totalRevenue": 150000.00,
                    "revenuePerUnit": 150.00,
                    "revenueGrowth": 12.5,
                    "averageOccupancy": 85.0
                },
                "costs": {
                    "operationalCosts": 45000.00,
                    "maintenanceCosts": 8000.00,
                    "securityCosts": 5000.00,
                    "staffCosts": 35000.00
                },
                "profitability": {
                    "grossMargin": 70.0,
                    "netMargin": 45.0,
                    "returnOnInvestment": 25.0,
                    "breakEvenPoint": 60.0
                },
                "billing": {
                    "paymentOnTime": 95.0,
                    "latePayments": 5.0,
                    "collectionRate": 98.0,
                    "averagePaymentTime": 2.5
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch financial KPIs: {str(e)}")

# ===== STORAGE MAP =====

@router.get("/map/{location_id}")
async def get_storage_map(location_id: str):
    """Get the interactive storage map for a location"""
    try:
        # TODO: Implement map data retrieval
        # Mock response for now
        return {
            "success": True,
            "data": {
                "locationId": location_id,
                "mapConfig": {
                    "width": 800,
                    "height": 600,
                    "gridSize": 50,
                    "snapToGrid": True,
                    "collisionDetection": True,
                    "undoRedo": True
                },
                "storageUnits": [
                    {
                        "id": "unit1",
                        "type": "POD",
                        "position": {"x": 100, "y": 200, "rotation": 0},
                        "status": "AVAILABLE"
                    }
                ],
                "zones": [
                    {
                        "id": "zone1",
                        "name": "Zone A",
                        "type": "POD",
                        "capacity": {
                            "totalUnits": 50,
                            "availableUnits": 35,
                            "utilizationRate": 70.0
                        }
                    }
                ],
                "capacity": {
                    "total": 100,
                    "available": 25,
                    "occupied": 75,
                    "utilizationRate": 75.0
                },
                "realTimeData": {
                    "lastUpdated": datetime.now().isoformat(),
                    "activeUsers": 5,
                    "recentChanges": [],
                    "alerts": []
                }
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch storage map: {str(e)}")

# ===== HEALTH CHECK =====

@router.get("/health")
async def storage_health_check():
    """Health check for storage system"""
    return {
        "success": True,
        "message": "Storage system is healthy",
        "timestamp": datetime.now().isoformat(),
        "version": "1.0.0"
    }
